File: A2Z_striver/1 Basics/1.2 Patterns/pattern_13/pattern_13.cpp
#include <iostream>

using namespace std;

// Header Comment Block
// Problem: Print Floyd's Triangle (Continuous Numbers).
// Example Input: n = 4
// Example Output:
// 1
// 2 3
// 4 5 6
// 7 8 9 10
// Goal: O(N^2) Time.

class Solution
{
public:
    // ~ Approach 1: The Accumulator (Standard)
    // Analysis: Time O(N^2) | Space O(1)
    // Pros: Simple, standard interview solution.
    void pattern13_accumulator(int n)
    {
        int num = 1;
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j <= i; j++)
            {
                cout << num << " ";
                num++;
            }
            cout << "\n";
        }
    }

    // ~ Approach 2: Mathematical / Stateless (Advanced)
    // Analysis: Time O(N^2) | Space O(1)
    // Pros: O(1) random access to any cell.
    void pattern13_stateless(int n)
    {
        for (int i = 0; i < n; i++)
        {

            // --- Math Logic Derivation ---
            // Question: What is the first number of Row 'i'?
            // Step 1: Count how many numbers exist in ALL previous rows (0 to i-1).
            //    - Row 0 has 1 number.
            //    - Row 1 has 2 numbers.
            //    - Row i-1 has 'i' numbers.
            //    - Total Previous Numbers = Sum of 1 to i = 1 + 2 + ... + i.

            // Step 2: Use the Gaussian Sum Formula for 'i' terms:
            //    - Sum = (i * (i + 1)) / 2

            // Step 3: The start of the current row is the NEXT number after that sum.
            //    - Start Value = Sum + 1

            int startVal = (i * (i + 1)) / 2 + 1;

            for (int j = 0; j <= i; j++)
            {
                // Current value is just Start + Offset (j)
                cout << (startVal + j) << " ";
            }
            cout << "\n";
        }
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    Solution sol;
    int n = 4;

    cout << "--- Approach 1: Accumulator ---\n";
    sol.pattern13_accumulator(n);

    cout << "\n--- Approach 2: Stateless (Math) ---\n";
    sol.pattern13_stateless(n);

    return 0;
}

File: A2Z_striver/1 Basics/1.2 Patterns/pattern_13/pattern_13.py
import sys

class Solution:
    """
    Problem: Print Floyd's Triangle (Continuous Numbers).
    Example Input: n = 4
    Example Output:
    1
    2 3
    4 5 6
    7 8 9 10
    Goal: O(N^2) Time.
    """

    # // ~ Approach 1: The Accumulator (Your Approach)
    # // Analysis: Time O(N^2) | Space O(1)
    # // Pros: Simple, keeps track of state naturally.
    def pattern13_accumulator(self, n):
        value = 1
        for i in range(n):
            for _ in range(i + 1):
                print(value, end=" ")
                value += 1
            print()

    # // ~ Approach 2: Stateless (Mathematical)
    # // Analysis: Time O(N^2) | Space O(1)
    # // Pros: Allows random access (calculating row 'i' without running previous rows).
    def pattern13_stateless(self, n):
        for i in range(n):
            # // --- Math Logic ---
            # // 1. Count total numbers in previous rows (0 to i-1).
            # //    Sum = 1 + 2 + ... + i = (i * (i+1)) // 2
            # // 2. Start of current row is Sum + 1.
            start_val = (i * (i + 1)) // 2 + 1
            
            for j in range(i + 1):
                # // Print StartVal + Offset (j)
                print(f"{start_val + j} ", end="")
            print()

# Driver Code
if __name__ == "__main__":
    solver = Solution()
    n = 4
    
    print(f"--- Approach 1: Accumulator (n={n}) ---")
    solver.pattern13_accumulator(n)
    
    print(f"\n--- Approach 2: Stateless (n={n}) ---")
    solver.pattern13_stateless(n)